taichi/ir/control_flow_graph.cpp

```diff
@@ -240,6 +240,15 @@ void CFGNode::reaching_definition_analysis(bool after_lower_access) {
     // loop in reversed order
     auto stmt = block->statements[i].get();
     auto data_source_ptrs = irpass::analysis::get_store_destination(stmt);
+    if (auto local_store = stmt->cast<LocalStoreStmt>()) {
```

Comment:
Note the special case here: this is because the granularity of reaching definition and liveness analysis is variable. This means the current implementation only tells us whether the entire matrix is dead or alive at a certain location. Therefore, if we make `get_store_destination` always return `ptr->origin`, during load-store forwarding optimization, a matrix could be replaced by the data loaded into it. For example, if we have
```
$1 = [a, b, c, d]
$2 = alloca
$3 = load [$2 <- $1]
$4 = Ptr shift [$3 + 0]
$5 = load [$4 <- x]
print $2
```
the last `$2` will be replaced by `x` if we directly change `get_store_destination`.